multicluster/controllers/multicluster/clusterset_controller.go

```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&multiclusterv1alpha1.ClusterSet{}).
 		Complete(r)
 }
+
+func (r *ClusterSetReconciler) validateLocalClusterClaim(clusterSet *multiclusterv1alpha1.ClusterSet) error {
+	// read the cluster claim in the cluster
+
+	configNamespace := clusterSet.GetNamespace()
+
+	clusterClaimList := &multiclusterv1alpha1.ClusterClaimList{}
+	r.Log.Info("Validating cluster claim in", "namespace", configNamespace)
+	if err := r.List(context.TODO(), clusterClaimList, client.InNamespace(configNamespace)); err != nil {
+		return err
+	}
+	if len(clusterClaimList.Items) == 0 {
+		return fmt.Errorf("cluster claim not configured for the cluster")
+	}
+
+	var clusterSetClaimID string
+	wellKnownClusterSetClaimIDExist := false
+	var clusterClaimID string
+	wellKnownClusterClaimIDExist := false
+	for _, clusterClaim := range clusterClaimList.Items {
+		r.Log.Info("Processing clusterclaim", "name", clusterClaim.Name, "value", clusterClaim.Value)
+		if strings.Compare(clusterClaim.Name, multiclusterv1alpha1.WellKnownClusterClaimClusterSet) == 0 {
+			wellKnownClusterSetClaimIDExist = true
+			clusterSetClaimID = clusterClaim.Value
+		} else if strings.Compare(clusterClaim.Name, multiclusterv1alpha1.WellKnownClusterClaimID) == 0 {
+			wellKnownClusterClaimIDExist = true
+			clusterClaimID = clusterClaim.Value
+		}
+	}
+
+	if !wellKnownClusterSetClaimIDExist {
+		return fmt.Errorf("clusterset claim ID not configured for the cluster")
+	}
+
+	if !wellKnownClusterClaimIDExist {
+		return fmt.Errorf("cluster claim ID not configured for the cluster")
+	}
+
+	configExists := false
+	if r.IsMcsLeader {
+		for _, leader := range clusterSet.Spec.Leaders {
+			if strings.Compare(clusterClaimID, leader.ClusterID) == 0 {
+				configExists = true
+			}
+		}
+		if !configExists {
+			return fmt.Errorf("cluster not defined as leader in cluster set")
+		}
+		// also validate the namespace is the same
+		if strings.Compare(clusterSet.Spec.Namespace, clusterSet.GetNamespace()) != 0 {
+			return fmt.Errorf("cluster set namespace " + clusterSet.Spec.Namespace + " is different from " +
+				clusterSet.GetNamespace())
+		}
+	} else {
+		for _, member := range clusterSet.Spec.Members {
+			if strings.Compare(clusterClaimID, member.ClusterID) == 0 {
+				configExists = true
+			}
+		}
+		if !configExists {
+			return fmt.Errorf("cluster not defined as member in cluster set")
+		}
+	}
+
+	r.clusterID = common.ClusterID(clusterClaimID)
+	r.clusterSetID = common.ClusterSetID(clusterSetClaimID)
+	r.clusterSetConfig = clusterSet.DeepCopy()
+
+	return nil
+}
+
+func (r *ClusterSetReconciler) createOrUpdateMultiClusterSetOnLeaderCluster(clusterSet *multiclusterv1alpha1.ClusterSet) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.inboundLeaderMgr == nil {
+		r.inboundLeaderMgr = internal.NewInboundLeaderClusterManager(r.Client, r.clusterID, clusterSet.GetNamespace(), r.Log)
```

Comment:
Could you help to check https://github.com/abhiraut/antrea/blob/63d4af9be579bfb9f09b55a354e541a76f13b778/multicluster/controllers/multicluster/resourceexport_controller.go#L56 or refer to Antrea code to provide a method like `NewClusterSetReconciler` to do initialization in one place?

Reply:
Ill add a todo and get to this in the next set of changes